### PR TITLE
Remove closure base file dependency

### PR DESCRIPTION
**Pull request by samelhusseini**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

Shave off **28KB** from Blockly compressed.

### Proposed Changes

Because we made use of a number of closure's base methods, we had a need to include closure's base file. Since we only use very few of these methods, we can trim down our binary size even further by providing alternatives for the following: 
- goog.inherits
- goog.exportSymbol
- goog.mixin
- goog.require
- goog.provide

``goog.exportSymbol`` and ``goog.mixin`` can be moved to Blockly.utils and re-implemented.

``goog.provide`` and ``goog.require`` are built in and the compiler knows how to handle those. We do need to stub those though, because our [message files](https://github.com/google/blockly/blob/master/msg/js/en.js#L7) include a require and provide call. 

``goog.inherits`` I've toyed with a few ideas of what to do with this one, but the least intrusive solution I came up with was to just inject an implementation of it.

@NeilFraser thoughts?

### Reason for Changes

Less closure (save 28KB)

### Test Coverage

Tested the demos.

### CODE REVIEW

File: core/utils.js [hunk 1]
```diff
@@ -622,3 +622,37 @@ Blockly.utils.screenToWsCoordinates = function(ws, screenCoordinates) {
...
 };
+
+/**
+ * Builds an object structure for the provided namespace path, ensuring that
+ * names that already exist are not overwritten. For example:
+ * "a.b.c" -> a = {};a.b={};a.b.c={}
+ * @param {string} name name of the object that this file defines.
+ * @param {*=} opt_object the object to expose at the end of the path.
+ */
+Blockly.utils.exportSymbol = function(name, opt_object) {
```
Comment (NeilFraser) [inline core/utils.js]: Find these comments in our codebase:
// Export symbols that would otherwise be renamed by Closure compiler.

And either replace them with exportSymbol, or delete this function and use the existing approach.  Whatever we do, it should be consistent.

File: build.py [hunk 2]
```diff
@@ -210,6 +210,10 @@ def gen_core(self):
         ("warning_level", "DEFAULT"),
       ]
 
+    # Add Closure provide, require and inherits methods.
+    params.append(("js_code", """var goog=goog||{};goog.provide=function(){};goog.require=function(){};
+goog.inherits=function(a,b){function c(){}c.prototype=b.prototype;a.superClass_=b.prototype;a.prototype=new c;a.prototype.constructor=a;};"""))
```
Comment (NeilFraser) [inline build.py]: Neat idea, but this moves some of our core code into a very unexpected location, build.py.  And it means everyone's build process needs a similar hack added.

File: build.py [hunk 3]
```diff
@@ -210,6 +210,10 @@ def gen_core(self):
         ("warning_level", "DEFAULT"),
       ]
 
+    # Add Closure provide, require and inherits methods.
+    params.append(("js_code", """var goog=goog||{};goog.provide=function(){};goog.require=function(){};
```
Comment (NeilFraser) [inline build.py]: I don't see any goog.provide or goog.require in the compressed files.  Why are these needed?

Comment (samelhusseini) [reply build.py]: It's there: https://github.com/google/blockly/blob/master/blockly_compressed.js#L7
We need it for our message files as they have provide/require calls eg: [here](https://github.com/google/blockly/blob/master/msg/js/en.js). and are included as is in html files eg: [here](https://github.com/google/blockly/blob/master/demos/fixed/index.html#L8)

Comment (samelhusseini) [reply build.py]: We also have calls to require/provide in our custom fields in demos, but I know we can get rid of those. I'm just not sure about the message files, does Blockly games need the call to require/provide in order to bundle message files?

File: core/utils.js (hunk 1 above)
Comment (samelhusseini) [reply core/utils.js]: Nice, done.

File: core/utils.js [hunk 4]
```diff
@@ -622,3 +622,37 @@ Blockly.utils.screenToWsCoordinates = function(ws, screenCoordinates) {
...
+      cur = cur[part] = {};
+    }
+  }
+};
+
+
+/**
+ * Copies all the members of a source object to a target object.
+ * @param {Object} target Target.
+ * @param {Object} source Source.
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply core/utils.js]: Done.

File: build.py (hunk 2 above)
Comment (samelhusseini) [reply build.py]: I'm happy moving these to a `base.min.js` file or something so they're on their own and under `closure/goog/`, but I can't think of a way to have them defined in `core/` and not have to:
- reimplement provide/require (as we need it in there for the reasons outlined in the PR description)
- have a need to depend (ie: require) on a "base" file anywhere we do a `goog.inherit`.
If we think of this patch as a way to make our code work with the closure compiler, then the build script is the appropriate place to patch them in.
For example, we have similar code [here](https://github.com/google/blockly/blob/master/build.py#L255) and [here](https://github.com/google/blockly/blob/master/build.py#L283).

File: core/utils.js [hunk 5]
```diff
@@ -622,3 +622,37 @@ Blockly.utils.screenToWsCoordinates = function(ws, screenCoordinates) {
   var finalOffsetMainWs = finalOffsetPixels.scale(1 / ws.scale);
   return finalOffsetMainWs;
 };
+
+/**
+ * Builds an object structure for the provided namespace path, ensuring that
+ * names that already exist are not overwritten. For example:
+ * "a.b.c" -> a = {};a.b={};a.b.c={}
+ * @param {string} name name of the object that this file defines.
+ * @param {*=} opt_object the object to expose at the end of the path.
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply core/utils.js]: Fixed.

File: gulpfile.js [hunk 6]
```diff
@@ -83,6 +83,14 @@ function prependHeader() {
   return gulp.insert.prepend(`// Do not edit this file; automatically generated by gulp.\n`);
 }
 
+/**
+ * Helper method for prepending Google closure methods used in Blockly.
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply gulpfile.js]: Fixed.

File: demos/custom-fields/pitch/field_pitch.js [hunk 7]
```diff
@@ -29,6 +29,7 @@ goog.provide('CustomFields.FieldPitch');
 goog.require('Blockly.FieldTextInput');
 goog.require('Blockly.utils.math');
 
+var CustomFields = {};
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply demos/custom-fields/pitch/field_pitch.js]: Fixed.

File: build.py (hunk 3 above)
Comment (samelhusseini) [reply build.py]: Discussed offline, we'll go ahead and remove the require/provide calls in the built message files.

File: core/workspace.js [hunk 8]
```diff
@@ -791,10 +791,11 @@ Blockly.Workspace.getAll = function() {
...
 
 // Export symbols that would otherwise be renamed by Closure compiler.
-Blockly.Workspace.prototype['clear'] = Blockly.Workspace.prototype.clear;
-Blockly.Workspace.prototype['clearUndo'] =
-    Blockly.Workspace.prototype.clearUndo;
-Blockly.Workspace.prototype['addChangeListener'] =
-    Blockly.Workspace.prototype.addChangeListener;
-Blockly.Workspace.prototype['removeChangeListener'] =
-    Blockly.Workspace.prototype.removeChangeListener;
+Blockly.utils.exportSymbol('Blockly.Workspace.prototype.clear',
```
[the remark this answers is not in the document]
Comment (samelhusseini) [reply core/workspace.js]: Discussed offline, will use @export instead.

File: build.py (hunk 2 above)
Comment (samelhusseini) [reply build.py]: Going to go ahead with Blockly.utils.object.inherits....

**Comment by samelhusseini** [comment]
@NeilFraser this is ready for a final review.

File: tests/compile/main.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 goog.provide('Main');
-// Messages (in some language)
-goog.require('Blockly.Msg.en');
+// Messages
+goog.require('Blockly.Msg');
```
Comment (NeilFraser) [inline tests/compile/main.js]: Why is this needed?

Comment (samelhusseini) [reply tests/compile/main.js]: Yep, removed.
